### PR TITLE
Add hex conversion SQL functions

### PR DESCRIPTION
**Pull request by zwass**
Provide functions for converting to and from hex strings.

### CODE REVIEW

**Comment by obelisk** [comment]
This needs a description but also how is it different than using this:
```
➜  osquery git:(master) ✗ osqueryi
Using a virtual database. Need help, type '.help'
osquery> select printf('%08X', 255);
+---------------------+
| printf('%08X', 255) |
+---------------------+
| 000000FF            |
+---------------------+
osquery>
```

**Comment by zwass** [comment]
The example you provide allows conversion from integer to hex string, but I don't know of any current way to convert from hex string to integer. Though there is this `printf` method for one direction of the conversion, I feel it is a much simpler UX to provide roundtrip functions for conversion.

File: osquery/sql/sqlite_string.cpp [hunk 1]
```diff
@@ -152,6 +152,41 @@ static void ip4StringToDecimalFunc(sqlite3_context* context,
   }
 }
 
+/**
+ * @brief Convert a hexadecimal string to integer
+ */
+static void hexStringToDecimalFunc(sqlite3_context* context,
+                                   int argc,
+                                   sqlite3_value** argv) {
+  assert(argc == 1);
```
Comment (obelisk) [inline osquery/sql/sqlite_string.cpp]: Did you leave this intentionally?

File: osquery/sql/sqlite_string.cpp [hunk 2]
```diff
@@ -152,6 +152,41 @@ static void ip4StringToDecimalFunc(sqlite3_context* context,
...
+
+  if (SQLITE_NULL == sqlite3_value_type(argv[0])) {
+    sqlite3_result_null(context);
+    return;
+  }
+
+  long value = std::strtoll((char*)sqlite3_value_text(argv[0]), nullptr, 16);
```
Comment (obelisk) [inline osquery/sql/sqlite_string.cpp]: tryTo<...> generics here instead of strtoll

File: osquery/sql/sqlite_string.cpp [hunk 3]
```diff
@@ -152,6 +152,41 @@ static void ip4StringToDecimalFunc(sqlite3_context* context,
...
+  sqlite3_result_int64(context, value);
+}
+
+/**
+ * @brief Convert an integer to hexadecimal string
+ */
+static void decimalToHexStringFunc(sqlite3_context* context,
+                                   int argc,
+                                   sqlite3_value** argv) {
+  assert(argc == 1);
```
Comment (obelisk) [inline osquery/sql/sqlite_string.cpp]: Did you leave this intentionally?

File: osquery/sql/sqlite_string.cpp [hunk 4]
```diff
@@ -152,6 +152,41 @@ static void ip4StringToDecimalFunc(sqlite3_context* context,
...
+
+  if (SQLITE_NULL == sqlite3_value_type(argv[0])) {
+    sqlite3_result_null(context);
+    return;
+  }
+
+  char str[65];
+  std::snprintf(str, 19, "%#x", sqlite3_value_int(argv[0]));
+  sqlite3_result_text(context, str, strlen(str), SQLITE_TRANSIENT);
```
Comment (obelisk) [inline osquery/sql/sqlite_string.cpp]: Should this be strnlen?

File: osquery/sql/sqlite_string.cpp (hunk 1 above)
Comment (zwass) [reply osquery/sql/sqlite_string.cpp]: Copied style from functions above.

File: osquery/sql/sqlite_string.cpp (hunk 4 above)
Comment (zwass) [reply osquery/sql/sqlite_string.cpp]: I can actually get the length from `snprintf` so I'll do that.

File: osquery/sql/sqlite_string.cpp (hunk 1 above)
Comment (zwass) [reply osquery/sql/sqlite_string.cpp]: Do you want me to remove this and/or the other asserts in this file?

**Comment by zwass** [comment]
Rebased on `experimental`. Can someone please take a look?

**Comment by zwass** [comment]
Ping. Can anyone take a look at this small PR?

**Comment by packetzero** [comment]
Looks fine to me.  I would prefer an optional second arg to to_hex() that will turn off '0x' prefix.

You should consider calling sqlite3_result_error() in cases where argv[0] is null, argv[0] has more than 16 hex chars (larger than int64).  Consider adding from_hex() tests with inputs:
'0xDEADBEEF123', '' , null, '0x0AAAABBBBCCCCDDDD', 'blah blah'

Optional paranoid tests with inputs: no args, 3 args, '0', '0x0', '0x'

Thanks!

**Comment by directionless** [comment]
I keep finding myself wanting other number bases. (octal, and binary specifically). What do you think about making a general conversion routine that takes the decimal and base as arguments?

**Comment by theopolis** [comment]
We'll need to rebase this

**Comment by Smjert** [comment]
ping @zwass, this needs a rebase to resolve conflicts.

**Comment by Smjert** [comment]
Ping again @zwass any time to work on this? 
Otherwise I suggest to close it and maybe open an issue about it and discuss also on the other number bases conversions.

If you have time to work on this: It needs to be rebased on latest master.

My two cents on the specifics of this PR:
I agree with @directionless, that it would be useful to have a generic function to do any kind of conversions.
Though we can also have both, in the sense that "to/from_hex", "to/from_oct" etc could be easy frontends to the generic base co nversion function.

**Comment by theopolis** [comment]
Heads up, we are going to close older PRs to help with project organization and focus. If you are still eager to land your changes please feel welcomed to re-open the request. If you feel like you were waiting for answers or a review from other project contributors please mention me and I'll take a look as soon as possible.

**Comment by directionless** [comment]
FWIW I don't remember how far I got it, but I did some tinkering in https://github.com/osquery/osquery/compare/master...directionless:seph/hex-functions

**Pull request closed without merging**